# Project Assignment 5: Data Aggregation and Group Operations

**Objective:**
In the previous assignment, you formatted and reshaped your data. Now, you will aggregate your dataset to compute summary statistics and analyze relationships between categorical variables.

**Guidelines:**
* **Mandatory Completion:** You may not need to perform all of the tasks listed below if your dataset lacks the required data types (e.g., categorical variables). However, you must not simply omit the task from the submission. 
* **Universal Justification:** For every section, you must write code or provide an explanation.
    * If you performed an operation, explain the resulting metric and its relevance to your project.
    * If you skipped an operation, use a markdown cell or code comment to state exactly why the task is not relevant to your specific dataset.
* **Reference:** Use the lecture notebook and the hands-on exercises to guide your implementation.

In [1]:
# --- Import Libraries ---
import pandas as pd
import numpy as np

# TODO: Load your processed dataset (the CSV you saved at the end of the previous assignment)
df_clean = pd.read_csv('project_data_processed.csv')
df_clean.head()
df_clean.info()
df_clean.describe(include="all")

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   order_id              50000 non-null  int64  
 1   order_date            50000 non-null  str    
 2   product_id            50000 non-null  int64  
 3   product_category      50000 non-null  str    
 4   price                 50000 non-null  float64
 5   discount_percent      50000 non-null  int64  
 6   quantity_sold         50000 non-null  int64  
 7   customer_region       50000 non-null  str    
 8   payment_method        50000 non-null  str    
 9   rating                50000 non-null  float64
 10  review_count          50000 non-null  int64  
 11  discounted_price      50000 non-null  float64
 12  total_revenue         50000 non-null  float64
 13  order_month           50000 non-null  int64  
 14  payment_method_clean  50000 non-null  str    
dtypes: float64(4), int64(6), str(5

,order_id,order_date,product_id,product_category,price,discount_percent,quantity_sold,customer_region,payment_method,rating,review_count,discounted_price,total_revenue,order_month,payment_method_clean
count,50000.000000,50000,50000.000000,50000,50000.000000,50000.000000,50000.000000,50000,50000,50000.000000,50000.000000,50000.000000,50000.000000,50000.0000,50000
unique,NaN,730,NaN,6,NaN,NaN,NaN,4,5,NaN,NaN,NaN,NaN,NaN,5
top,NaN,2023-10-19,NaN,Beauty,NaN,NaN,NaN,Asia,Wallet,NaN,NaN,NaN,NaN,NaN,Wallet
freq,NaN,93,NaN,8465,NaN,NaN,NaN,12526,10106,NaN,NaN,NaN,NaN,NaN,10106
mean,25000.500000,NaN,2986.848740,NaN,252.507260,13.340700,2.999400,NaN,NaN,2.996316,249.329280,218.886566,657.331475,6.5044,NaN
std,14433.901067,NaN,1156.374535,NaN,143.025544,9.850694,1.415401,NaN,NaN,1.154295,144.251981,127.317681,526.223968,3.4360,NaN
min,1.000000,NaN,1000.000000,NaN,5.010000,0.000000,1.000000,NaN,NaN,1.000000,0.000000,3.530000,3.840000,1.0000,NaN
25%,12500.750000,NaN,1983.000000,NaN,127.840000,5.000000,2.000000,NaN,NaN,2.000000,125.000000,109.680000,240.320000,4.0000,NaN
50%,25000.500000,NaN,2983.000000,NaN,252.970000,10.000000,3.000000,NaN,NaN,3.000000,250.000000,215.805000,505.410000,7.0000,NaN
75%,37500.250000,NaN,3989.000000,NaN,376.335000,20.000000,4.000000,NaN,NaN,4.000000,374.000000,322.702500,968.970000,9.0000,NaN


## Part 1: Grouping Data (1.5 Points)

Grouping allows you to segment your data and compute metrics (like mean, sum, or count) for each segment.

**Your Task:**
1.  Identify a categorical variable to group by.
2.  Apply an aggregation function to one or more numerical columns.
3.  If your dataset does not support grouping, explain the limitation in the markdown cell below.

In [2]:
# TODO: Perform a groupby operation
# Example: df_grouped = df.groupby('category_column')['numeric_column'].mean()

# TODO: Display the result
# print(df_grouped.head())
category_summary = df_clean.groupby("product_category").agg({
    "total_revenue": "sum",
    "quantity_sold": "sum",
    "price": "mean",
    "rating": "mean"
}).reset_index()

category_summary

,product_category,total_revenue,quantity_sold,price,rating
0,Beauty,5550624.97,25422,252.405289,2.985186
1,Books,5484863.03,25065,252.682468,3.020259
2,Electronics,5470594.03,24898,251.893404,2.991298
3,Fashion,5480123.34,25089,252.347879,2.987782
4,Home & Kitchen,5473132.55,24743,253.812631,2.996706
5,Sports,5407235.82,24753,251.910161,2.996891


**Reflection:**
*Double click here to edit.*

I grouped the dataset by the categorical variable product_category and calculated several summary statistics for each category, including total revenue, total quantity sold, average price, and average rating. This aggregation summarizes overall performance within each product category rather than looking at individual transactions.

The grouped metrics provide insight into how different categories contribute to overall sales. For example, the results show that categories such as Beauty, Books, and Fashion generate similar levels of total revenue, each around 5.4–5.5 million. The quantity sold is also relatively similar across categories, suggesting that demand is fairly evenly distributed among product types in the dataset.

The average price and rating help provide additional context. Prices are fairly consistent across categories, averaging around $252, while ratings are close to 3 across all groups. This indicates that there are no large differences in pricing or customer satisfaction between product categories.

For the final analysis, this grouped information helps identify whether certain categories dominate revenue or sales and whether product characteristics differ across categories. It provides a higher level view of product performance that can guide further analysis, such as exploring how discounts, regions, or payment methods influence sales within specific categories.

## Part 2: Pivot Tables (1.5 Points)

Pivot tables provide a multi-dimensional summary of your data, allowing you to calculate metrics across two or more categorical axes.

**Your Task:**
1.  Construct a pivot table using `pd.pivot_table()`.
2.  Define the `index`, `columns`, and `values` parameters.
3.  If your dataset does not have suitable dimensions for a pivot table, explain the omission below.

In [3]:
# TODO: Create a pivot table
# df_pivot = pd.pivot_table(df, values='numeric_val', index='category_1', columns='category_2', aggfunc='sum')

# TODO: Display the result
# print(df_pivot.head())
df_pivot = pd.pivot_table(
    df_clean,
    values="total_revenue",
    index="product_category",
    columns="customer_region",
    aggfunc="sum"
)

print(df_pivot)

customer_region         Asia      Europe  Middle East  North America
product_category                                                    
Beauty            1401916.92  1358226.42   1387711.96     1402769.67
Books             1391961.69  1330671.45   1428560.07     1333669.82
Electronics       1319074.46  1407118.94   1379642.77     1364757.86
Fashion           1334485.23  1366107.00   1390228.60     1389302.51
Home & Kitchen    1369676.45  1326424.56   1375546.52     1401485.02
Sports            1358085.08  1323763.20   1340154.58     1385232.96


**Reflection & Justification:**
*Double click here to edit.*

I created a pivot table using product_category as the index and customer_region as the columns, with total_revenue as the aggregated value. This structure allows the table to summarize how revenue is distributed across both product types and geographic regions simultaneously. Each row represents a product category, and each column represents a region. The values show the total revenue generated for that specific category–region combination.

This pivot table reveals how sales performance varies across markets. Overall, revenue is relatively evenly distributed across regions, with Asia, Europe, the Middle East, and North America each contributing similar amounts of revenue for most categories. However, small differences still appear. For example, Electronics generates slightly higher revenue in Europe, while Beauty and Home & Kitchen show strong revenue in North America. Books perform particularly well in the Middle East compared to other regions.

The table exposes the interaction between product demand and geographic markets, which would not be as clear using a single grouping variable. This type of summary is useful for identifying regional demand patterns and determining whether certain product categories perform better in specific areas. In the context of the final analysis, this helps highlight potential regional preferences and provides insight into where companies might focus marketing or inventory decisions.

## Part 3: Cross-Tabulation (1.0 Point)

Cross-tabulation is highly effective for computing simple cross-frequencies of two or more categorical variables.

**Your Task:**
1.  Use `pd.crosstab()` to analyze the frequency distribution between categorical features.
2.  Alternatively, provide the required justification if this task is omitted.

In [4]:
# TODO: Generate a cross-tabulation
# df_cross = pd.crosstab(df['category_A'], df['category_B'])

# TODO: Display the result
# print(df_cross.head())
df_cross = pd.crosstab(df_clean['product_category'], df_clean['payment_method_clean'])

print(df_cross)

payment_method_clean  Cash On Delivery  Credit Card  Debit Card   Upi  Wallet
product_category                                                             
Beauty                            1751         1639        1696  1659    1720
Books                             1669         1685        1629  1696    1648
Electronics                       1649         1687        1669  1676    1639
Fashion                           1608         1635        1667  1720    1735
Home & Kitchen                    1622         1587        1678  1663    1708
Sports                            1628         1675        1642  1664    1656


**Reflection & Justification:**
*Double click here to edit.*

I generated a cross tabulation between product_category and payment_method_clean to analyze the frequency distribution of payment methods across different product categories. In this table, each row represents a product category and each column represents a payment method. The values show the number of orders made using each payment method within each category.

The frequencies indicate that payment methods are fairly evenly distributed across categories, with no single payment type dominating purchases for any specific category. For example, Beauty, Books, Electronics, and other categories all show similar counts across payment methods such as Credit Card, Debit Card, UPI, Wallet, and Cash on Delivery. This suggests that customer payment preferences do not vary dramatically depending on the type of product purchased.

This cross tabulation helps provide a clearer understanding of purchasing behavior within the dataset. It shows that payment method usage is relatively balanced, indicating that customers across different product categories rely on a mix of payment options. For the final analysis, this suggests that payment method choice is likely not strongly tied to product category, but it still provides useful insight into the overall distribution of transaction types in the dataset.

## Part 4: Final Reflection on your Research Question (1.0 Point)

**Implications for your research questions:**
*Double-click here to edit.*

State how the findings from your data aggregations and group operations inform the research questions you proposed in Project Assignment 1 and Lab 5. Cite specific numerical results from the tables generated above to support your statements. Please indicate how these statistical summaries support your initial analytical intent or whether you might need to update your approach and how.

**Examples of valid reflections:**
Implications for your research questions

Which product categories perform best?
The grouped results from Part 1 summarize performance by product_category using total revenue and quantity sold. The table shows that Beauty generates the highest total revenue at 5,550,624.97 with 25,422 units sold, followed by Books with 5,484,863.03 and 25,065 units sold, and Fashion with 5,480,123.34 and 25,089 units sold. The lowest performing category in terms of revenue is Sports with 5,407,235.82 and 24,753 units sold.

Although Beauty slightly outperforms the other categories, the revenue and sales volumes across all categories are relatively similar, each generating around 5.4–5.5 million in revenue and roughly 24,700–25,400 units sold. This suggests that demand is fairly evenly distributed across product categories rather than concentrated in one dominant category. These findings support the initial analytical goal of identifying top performing categories, while also indicating that category alone may not strongly explain variation in sales performance.

Are ratings and review counts linked to higher sales?
The aggregated statistics from Part 1 also show that the average rating across all product categories is very similar, ranging from about 2.98 to 3.02. For example, Books have the highest average rating at 3.02, while Beauty has an average rating of 2.99. Despite Books having the highest rating, it does not generate the highest revenue, while Beauty produces the most revenue despite having a slightly lower average rating.

This suggests that higher ratings alone may not directly translate into higher revenue or sales volume in this dataset. Because ratings are very similar across categories, other factors such as price, discounts, or quantity sold may play a larger role in explaining sales outcomes. Therefore, for the final analysis it may be necessary to examine the combined relationship between ratings, review counts, and sales metrics rather than relying on ratings alone to explain product performance